decision_service/dsclientlib/logger/ds_async_batcher.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "ds_concurrent_queue.h"
+#include "ds_api_status.h"
+
+#include <memory>
+#include <string>
+#include "../error_callback_fn.h"
+#include <thread>
+
+namespace decision_service {
+  class error_callback_fn;
+
+	// This class takes uses a queue and a background thread to accumulate events, and send them by batch asynchronously.
+	// A batch is shipped with TSender::send(data)
+	template <typename TSender>
+	class async_batcher {
+
+	public:
+
+		int append(const std::string& evt, api_status* status = nullptr)
+		{
+			if (_queue.size() < _queue_max_size) {
+				_queue.push(evt);
+				return error_code::success;
+			}
+
+			//report errors
+			api_status::try_update(status, error_code::background_queue_overflow, "dropped event: " + evt);
+			return error_code::background_queue_overflow;
+		}
+
+	private:
+		void timer()//the timer triggers a queue flush (run in background)
+		{
+			while (_thread_is_running)
+			{
+				std::this_thread::sleep_for(std::chrono::milliseconds(_batch_timeout_ms));
+				flush();
+			}
+		}
+
+		void flush()//flush all batches
+		{
+		  const auto queue_size = _queue.size();
+			if (queue_size == 0) return;
+
+			//handle batching
+			std::string batch, next;
+			_queue.pop(&batch); //there is at least one element
+
+			for (size_t i = 1; i < queue_size; ++i)
+			{
+				_queue.pop(&next);
+
+				size_t batch_size = batch.length() + next.length();
+				if (batch_size > _batch_max_size)
+				{
+          api_status status;
+					//the batch is about to reach the max size: send it
+					if(_sender.send(batch,&status) != error_code::success)
+					{
+            if (_perror_cb) _perror_cb->report_error(status);
```

Comment:
> if (_perror_cb) _perror_cb->report_error(status); [](start = 11, length = 50)

maybe introduce a try_report_error function similar to api_status?

Reply:
maybe introduce a similar macro to REPORT for callback.

What to do if there is a stack of call? call at each level? call at the end?

---
In reply to: [186743495](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186743495) [](ancestors = 186743495)